db/db_impl_write.cc

```diff
@@ -1281,7 +1281,9 @@ Status DBImpl::SwitchMemtable(ColumnFamilyData* cfd, WriteContext* context,
   // In case of pipelined write is enabled, wait for all pending memtable
   // writers.
   if (immutable_db_options_.enable_pipelined_write) {
+    mutex_.Unlock();
```

Comment:
I would prefer passing `mutex_` into `WaitForMemTableWriters` and only unlock and re-lock mutex around `AwaitState`: https://github.com/facebook/rocksdb/blob/master/db/write_thread.cc#L664

Reply:
WriteThread::EnterUnbatched   first unlock, then call WaitForMemTableWriters, last  lock. if modify it as you said,  it has problem

Reply:
What I mean was changing signature of WaitForMemTableWriters() into:
```
    void WaitForMemTableWriters(InstrumentedMutex* mutex = nullptr);
```
If mutex is nullptr, the unlock mechanism is skipped. We can keep other callers unchanged and only pass in mutex from this line. What I want to do was avoid unnecessary unlocking. But after a second thought since SwithMemtable() is not in a critical path I think the current patch is fine.

btw do you mind adding some inline comment here explain about the deadlock?

Reply:
thread 1  finish write wal, and writing memtable.  data type is  kTypeMerge and the number of successive merges >= max_successive_merges, so thread 1 will get data.  GetImpl call  GetAndRefSuperVersion, sometimes it need lock.  
thread 2 call PipelinedWriteImpl, then it lock, it call PreprocessWrite, it find flush_scheduler_ is not empty. so thread 2 call ScheduleFlushes,  then call SwitchMemtable, then found enable_pipelined_write = true, so it will call WaitForMemTableWriters, thus it will wait signal.

Reply:
thanks, but I mean adding brief inline comment in the code like
```
     // Memtable writers may call DB::Get in case max_successive_merges > 0, which may lock mutex. Unlocking mutex here to avoid deadlock.
```